presto-main/src/main/java/com/facebook/presto/metadata/Signature.java

```diff
@@ -328,15 +328,22 @@ private static boolean matchAndBind(Map<String, Type> boundParameters, Map<Strin
     {
         // If this parameter is already bound, then match (with coercion)
         if (boundParameters.containsKey(parameter.getBase())) {
-            checkArgument(parameter.getParameters().isEmpty(), "Unexpected parameteric type");
+            checkArgument(parameter.getParameters().isEmpty(), "Unexpected parametric type");
             if (allowCoercion) {
                 if (canCoerce(type, boundParameters.get(parameter.getBase()))) {
                     return true;
                 }
                 else if (canCoerce(boundParameters.get(parameter.getBase()), type) && typeParameters.get(parameter.getBase()).canBind(type)) {
-                    // Broaden the binding
+                    // Try to coerce current binding to new candidate
                     boundParameters.put(parameter.getBase(), type);
                     return true;
+                } else {
+                    // Try to use common super type of current binding and candidate
```

Comment:
I agree that supertype may mean different things in different context.
I do not quite get why only same base type should be considered.

I can leave it as is. Or for now make it work only for DECIMAL type. 
But it should probably be possible to add (SHORTINT and DECIMAL(10)).
Then I opt for leaving this as is.